### PR TITLE
Add the ability to specify unlimited for individual resources

### PR DESCRIPTION
**Pull request by tylerslaton**
This adds the ability for `Unlimited` to be more granularly set per resource.

### Checklist
- [ ] The title of this PR would make a good line in Acorn's Release Note's Changelog
- [ ] The title of this PR ends with a link to the main issue being address in parentheses, like: `This is a title (#1216)`. [Here's an example](https://github.com/acorn-io/runtime/pull/1199)
- [ ] All relevant issues are referenced in the PR description. *NOTE: don't use [GitHub keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) that auto-close issues*
- [ ] Commits follow [contributing guidance](https://github.com/acorn-io/runtime/blob/main/CONTRIBUTING.md#commits)
- [ ] Automated tests added to cover the changes. If tests couldn't be added, an explanation is provided in the Verification and Testing section
- [ ] Changes to user-facing functionality, API, CLI, and upgrade impacts are clearly called out in PR description
- [ ] PR has at least two approvals before merging (or a reasonable exception, like it's just a docs change)

### CODE REVIEW

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 1]
```diff
@@ -84,14 +91,14 @@ func (current *Resources) Fits(incoming Resources) error {
...
-		if currentVal < incomingVal {
+		if currentVal != Unlimited && currentVal < incomingVal {
 			exceededResources = append(exceededResources, resource)
 		}
 	}
 
 	// Define function for checking quantity resources to keep code DRY
 	checkQuantityResource := func(resource string, currentVal, incomingVal resource.Quantity) {
-		if currentVal.Cmp(incomingVal) < 0 {
+		if !currentVal.Equal(UnlimitedQuantity()) && currentVal.Cmp(incomingVal) < 0 {
```
Comment (thedadams) [inline pkg/apis/internal.admin.acorn.io/v1/resources.go:101]: Personally, I would like to find a way to get rid of `UnlimiteQuantity`. Can `CmpInt64(Unlimited)` be used here instead?

Comment (tylerslaton) [reply pkg/apis/internal.admin.acorn.io/v1/resources.go:101]: Sure, I can do that.

Do you want to get rid of the `UnlimitedQuantity` function entirely? I mainly made it to make setting unlimited quantities more straight forward. Without it there would be a lot of `resource.NewQuantity(Unlimited,  resource.DecimalSI)` set. Is that something we'd want as it is more explicit?

Comment (thedadams) [reply pkg/apis/internal.admin.acorn.io/v1/resources.go:101]: No, I think it's fine then. Thanks for the explanation.

**Review by thedadams — APPROVED** [review]

**Review by keyallis — APPROVED** [review]

**Comment by tylerslaton** [comment]
@keyallis @thedadams There was an edge case the original code you reviewed did not account for. I put it into a separate commit with some unit tests. Please let me know what you think.

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 2]
```diff
@@ -30,44 +37,78 @@ type Resources struct {
...
 func (current *Resources) Remove(incoming Resources, all bool) {
-	// Do not allow resources to go below 0
-	nonNegativeSubtract := func(currentVal, incomingVal int) int {
-		difference := currentVal - incomingVal
+	sub := func(c, i int) int {
+		if c == Unlimited && i == Unlimited {
+			return 0
+		} else if c == Unlimited || i == Unlimited {
+			return c
+		}
```
Comment (keyallis) [inline pkg/apis/internal.admin.acorn.io/v1/resources.go:75]: I think your current code has a bug if `c` is some value and `i` is `Unlimited`.

Cases:
1. `c unlimited i unlimited -> 0`
1. `c unlimited i value -> unlimited`
1. `c value i unlimited -> 0`
1. `c value i value -> fallthrough case`

Here's the logic I came up with:
```
if i == Unlimited {
   return 0
} else if c == Unlimited {
   return c
}
```

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 3]
```diff
@@ -30,44 +37,78 @@ type Resources struct {
...
-	current.Jobs = nonNegativeSubtract(current.Jobs, incoming.Jobs)
-	current.Volumes = nonNegativeSubtract(current.Volumes, incoming.Volumes)
-	current.Images = nonNegativeSubtract(current.Images, incoming.Images)
-	current.Projects = nonNegativeSubtract(current.Projects, incoming.Projects)
+	subQuantity := func(c, i resource.Quantity) resource.Quantity {
+		if c.Equal(UnlimitedQuantity()) && i.Equal(UnlimitedQuantity()) {
+			return *resource.NewQuantity(0, c.Format)
+		} else if c.Equal(UnlimitedQuantity()) || i.Equal(UnlimitedQuantity()) {
+			return c
+		}
```
Comment (keyallis) [inline pkg/apis/internal.admin.acorn.io/v1/resources.go:89]: Same comment as above

File: pkg/apis/internal.admin.acorn.io/v1/resources_test.go [hunk 4]
```diff
@@ -83,44 +129,94 @@ func TestRemove(t *testing.T) {
...
+			},
+			incoming: Resources{
+				Apps:   Unlimited,
+				Memory: UnlimitedQuantity(),
+			},
+			expected: Resources{
+				Apps:   1,
+				Memory: resource.MustParse("1Mi"),
+			},
+		},
```
Comment (keyallis) [inline pkg/apis/internal.admin.acorn.io/v1/resources_test.go:206]: This is the test case where we would want to determine if the expected should be zeroed out or just left as current

**Review by keyallis — APPROVED** [review]

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 5]
```diff
@@ -9,6 +9,13 @@ import (
 
 var ErrExceededResources = fmt.Errorf("quota would be exceeded for resources")
 
+const Unlimited = -1
```
Comment (njhale) [inline pkg/apis/internal.admin.acorn.io/v1/resources.go:12]: Does this need to be exported? Is it used elsewhere? I'm assuming the answer is "yes", but figured I'd ask anyway.

**Review by njhale — APPROVED** [review]
One question, but not blocking.

LGTM

File: pkg/apis/internal.admin.acorn.io/v1/resources.go (hunk 5 above)
Comment (tylerslaton) [reply pkg/apis/internal.admin.acorn.io/v1/resources.go:12]: Yep, it will be used elsewhere, namely in Manager.

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 6]
```diff
@@ -30,44 +37,75 @@ type Resources struct {
...
-// Remove will remove the resources of another Resources struct from the current one.
+// Remove will remove the resources of another Resources struct from the current one. Calling remove
+// will be a no-op for any resource values that are set to unlimited.
 func (current *Resources) Remove(incoming Resources, all bool) {
-	// Do not allow resources to go below 0
-	nonNegativeSubtract := func(currentVal, incomingVal int) int {
-		difference := currentVal - incomingVal
+	sub := func(c, i int) int {
+		if c == Unlimited || i == Unlimited {
+			return c
```
Comment (thedadams) [inline pkg/apis/internal.admin.acorn.io/v1/resources.go:73]: Should this be
```suggestion
			return Unlimited
```

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 7]
```diff
@@ -30,44 +37,75 @@ type Resources struct {
...
 
-	current.Apps = nonNegativeSubtract(current.Apps, incoming.Apps)
-	current.Containers = nonNegativeSubtract(current.Containers, incoming.Containers)
-	current.Jobs = nonNegativeSubtract(current.Jobs, incoming.Jobs)
-	current.Volumes = nonNegativeSubtract(current.Volumes, incoming.Volumes)
-	current.Images = nonNegativeSubtract(current.Images, incoming.Images)
-	current.Projects = nonNegativeSubtract(current.Projects, incoming.Projects)
+	subQuantity := func(c, i resource.Quantity) resource.Quantity {
+		if c.Equal(UnlimitedQuantity()) || i.Equal(UnlimitedQuantity()) {
+			return c
```
Comment (thedadams) [inline pkg/apis/internal.admin.acorn.io/v1/resources.go:85]: Should this be
```suggestion
			return UnlimitedQuantity()
```

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 8]
```diff
@@ -30,44 +37,75 @@ type Resources struct {
...
+	add := func(c, i int) int {
+		if c == Unlimited || i == Unlimited {
+			return Unlimited
+		}
+		return c + i
+	}
+
+	addQuantity := func(c, i resource.Quantity) resource.Quantity {
+		if c.Equal(UnlimitedQuantity()) || i.Equal(UnlimitedQuantity()) {
+			return UnlimitedQuantity()
```
[the remark this answers is not in the document]
Comment (tylerslaton) [reply pkg/apis/internal.admin.acorn.io/v1/resources.go:49]: You're right, updating this code now.

File: pkg/apis/internal.admin.acorn.io/v1/resources.go (hunk 6 above)
Comment (thedadams) [reply pkg/apis/internal.admin.acorn.io/v1/resources.go:74]: @tylerslaton Just wanted to double-check whether this is correct. Same for the below.

Comment (tylerslaton) [reply pkg/apis/internal.admin.acorn.io/v1/resources.go:81]: Talked about this offline and left a comment to reflect why this code exists and what it is doing.

File: pkg/apis/internal.admin.acorn.io/v1/resources.go [hunk 9]
```diff
@@ -30,44 +35,88 @@ type Resources struct {
...
-	nonNegativeSubtract := func(currentVal, incomingVal int) int {
-		difference := currentVal - incomingVal
+	sub := func(c, i int) int {
+		// We don't expect this situation to happen. This is because there should not be a situation
+		// where we are removing from or with unlimited resources. However if it does, we want to
+		// be careful and handle it. With that in mind the logic here is as follows:
+		//
+		// 1. If the current value is unlimited, then removing a non-unlimited value should not change
+		//    the current value.
+		// 2. If the current value is not unlimited, then removing a non-unlimited value should not
```
Comment (thedadams) [inline pkg/apis/internal.admin.acorn.io/v1/resources.go:77]: nit:
```suggestion
		// 2. If the current value is not unlimited, then removing an unlimited value should not
```

**Review by thedadams — APPROVED** [review]
